### PR TITLE
Add a few easy module typing annotations

### PR DESCRIPTION
**Pull request by dcbaker**
This covers the test module and the icestorm module.

### CODE REVIEW

File: mesonbuild/modules/unstable_icestorm.py [hunk 1]
```diff
@@ -12,78 +12,119 @@
...
     @FeatureNew('FPGA/Icestorm Module', '0.45.0')
-    def __init__(self, interpreter):
+    def __init__(self, interpreter: Interpreter) -> None:
         super().__init__(interpreter)
-        self.yosys_bin = None
+        self.yosys_bin: T.Optional[ExternalProgram] = None
+        self.arachne_bin: T.Optional[ExternalProgram] = None
+        self.icepack_bin: T.Optional[ExternalProgram] = None
+        self.iceprog_bin: T.Optional[ExternalProgram] = None
+        self.icetime: T.Optional[ExternalProgram] = None
```
Comment (eli-schwartz) [inline mesonbuild/modules/unstable_icestorm.py:47]: In other modules we turned this into a `self.tools: T.Dict`.

Comment (dcbaker) [reply mesonbuild/modules/unstable_icestorm.py]: Yeah, that seems better.

File: mesonbuild/modules/unstable_icestorm.py [hunk 2]
```diff
@@ -12,78 +12,115 @@
...
+        proj_name, arg_sources = args
+        all_sources = self.interpreter.source_strings_to_files(
+            list(itertools.chain(arg_sources, kwargs['sources'])))
+
+        blif_target = build.CustomTarget(
+            f'{proj_name}_blif',
+            state.subdir,
+            state.subproject,
+            [self.tools['yosys'], '-q', '-p', 'synth_ice40 -blif @OUTPUT@', '@INPUT@'],
+            all_sources,
```
Comment (eli-schwartz) [inline mesonbuild/modules/unstable_icestorm.py:81]: mypy thinks that all_sources might contain StructuredSources here.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]
Could we re-order the 3rd commit to the beginning, so we don't add a bunch of self.X attributes and then immediately delete them?

File: mesonbuild/modules/unstable_icestorm.py (hunk 2 above)
Comment (dcbaker) [reply mesonbuild/modules/unstable_icestorm.py:81]: Yeah, I was unmotivated to tackle it

Comment (dcbaker) [reply mesonbuild/modules/unstable_icestorm.py:81]: I’ll deal with it Monday

Comment (eli-schwartz) [reply mesonbuild/modules/unstable_icestorm.py:81]: ping?

Comment (eli-schwartz) [reply mesonbuild/modules/unstable_icestorm.py:81]: Ping?

**Review by eli-schwartz — APPROVED** [review]

**Pull request merged**